### PR TITLE
fix: Encoding of List offsets was incorrect when slice offsets begin with zero

### LINKED ISSUE #6803: Arrow Flight Encodes a Slice's List Offsets If the slice offset is starts with zero
**Describe the bug**
If the arrow flight encodes a record batch slice where the first row of that slice has offset zero then it reuses the non-sliced data for offsets.  This can not only cause offset arrays that are larger than the slice to be encoded but the offset encoding will be incorrect as there may be zero or more rows that were not removed from the slice.

**To Reproduce**
```rust

    fn generate_nested_list_data_starting_at_zero<O: OffsetSizeTrait>() -> GenericListArray<O> {
        let mut ls =
            GenericListBuilder::<O, _>::new(GenericListBuilder::<O, _>::new(UInt32Builder::new()));


        for _i in 0..999 {
            ls.values().append(true);
            ls.append(true);
        }

        for j in 0..10 {
            for value in [j, j, j, j] {
                ls.values().values().append_value(value);
            }
            ls.values().append(true)
        }
        ls.append(true);


        for i in 0..9_000 {
            for j in 0..10 {
                for value in [i+j, i+j, i+j, i+j] {
                    ls.values().values().append_value(value);
                }
                ls.values().append(true)
            }
            ls.append(true);
        }

        ls.finish()
    }

    #[test]
    fn encode_nested_lists_starting_at_zero() {
        let inner_int = Arc::new(Field::new("item", DataType::UInt32, true));
        let inner_list_field = Arc::new(Field::new("item", DataType::List(inner_int), true));
        let list_field = Field::new("val", DataType::List(inner_list_field), true);
        let schema = Arc::new(Schema::new(vec![list_field]));

        let values = Arc::new(generate_nested_list_data_starting_at_zero::<i32>());

        let in_batch = RecordBatch::try_new(schema, vec![values]).unwrap();
        roundtrip_ensure_sliced_smaller(in_batch, 1);
    }
```

will result in an error where all lists are empty.

**Expected behavior**
No error is thrown at list offsets are properly encoded.

**Additional context**
This line seems to be the problem: https://github.com/apache/arrow-rs/blob/5a86db3c47d692584f694701780b0d8944a5e984/arrow-ipc/src/writer.rs#L1433
Setting that line to instead `0 => offset_slice.iter().map(|x| *x).collect(),` fixes the problem.

### PR DESCRIPTION
**Pull request by HawaiianSpork**
When encoding offsets the code had an optimization to reuse the offsets if the first offset was zero assuming the slice already pointed to first element. But the offset can also be zero if all previous lists were empty. When this occured it mold make all lists in the slice as empty, even if they shouldn't be.

# Which issue does this PR close?

Closes #6803 .

# Rationale for this change
 
Fixing a bug that can lead to loss of information when encoding record batches.

# What changes are included in this PR?

- Change the encoding and a test to demonstrate the fix.

# Are there any user-facing changes?

No, though some clients will experience smaller encoded messages since only the relevant slice will be sent.

### CODE REVIEW

File: arrow-ipc/src/writer.rs [hunk 1]
```diff
@@ -1429,8 +1429,8 @@ fn reencode_offsets<O: OffsetSizeTrait>(
     let start_offset = offset_slice.first().unwrap();
     let end_offset = offset_slice.last().unwrap();
 
-    let offsets = match start_offset.as_usize() {
-        0 => offsets.clone(),
+    let offsets: Buffer = match start_offset.as_usize() {
+        0 => offset_slice.iter().copied().collect(),
```
Comment (etseidl) [inline arrow-ipc/src/writer.rs:1433]: Would something like `0 => Buffer::from_slice_ref(offset_slice)` be better performing? Just curious as I'm still learning. 😄

File: arrow-ipc/src/writer.rs [hunk 2]
```diff
@@ -1429,8 +1429,8 @@ fn reencode_offsets<O: OffsetSizeTrait>(
     let start_offset = offset_slice.first().unwrap();
     let end_offset = offset_slice.last().unwrap();
 
-    let offsets = match start_offset.as_usize() {
-        0 => offsets.clone(),
+    let offsets: Buffer = match start_offset.as_usize() {
+        0 => Buffer::from_slice_ref(offset_slice),
```
Comment (tustvold) [inline arrow-ipc/src/writer.rs:1433]: Could we instead slice the Buffer passed in, this would avoid copying?

Comment (tustvold) [inline arrow-ipc/src/writer.rs:1433]: Something like

```suggestion
        0 => {
            let size = std::mem::sizeof::<O>();
            offsets.slice_with_length(data.offset()*size, (data.offset() + data.length() + 1) * size)
        }
```

File: arrow-ipc/src/writer.rs (hunk 1 above)
Comment (HawaiianSpork) [reply arrow-ipc/src/writer.rs]: Good call, that would be more efficient than the iterating approach.

File: arrow-ipc/src/writer.rs (hunk 2 above)
Comment (HawaiianSpork) [reply arrow-ipc/src/writer.rs]: Good recommendation.  code is changed.

**Comment by tustvold** [comment]
CI seems to be legitimately failing

**Comment by HawaiianSpork** [comment]
> CI seems to be legitimately failing

missed that.  fixed

**Comment by tustvold** [comment]
Looks good to me, thank you

**Pull request merged**